### ISSUE TITLE
drop proxy contract in favour of set_code_hash() and tagged storage

### ISSUE THREAD

**Issue by goastler** [prosopo/captcha#193]

**Comment by goastler** [comment]
tldr; Hernando said at the conference that it would be easier to tag our storage fields and manage the migration of storage manually than using some form of single proxy pointing at a contract which we redeploy for each update.

In a nutshell, the structure would be a single captcha contract with tagged storage fields. The update process would then be:
1. add / remove / change a storage field, say, or make some other edit to the contract code
2. make sure all fields are tagged and no duplicates (against historical tags also, they should **never** be reused)
3. build a fn in the contract to migrate any fields that have been removed into new/existing storage fields, make this only callable by admin
4. put the new contract code on chain
5. do a set_code_hash() on the current contract to switch the contract code to the new code
6. call the migrate fn from above to migrate old, deprecated storage over to new storage fields
7. rebuild the contract without the migrate fn, as we've now migrated and it is no longer needed
8. put this code on chain
9. call set_code_hash() to switch over to the new code except without the migrate fn

Perhaps steps 7-9 can be skipped as the migrate fn existing isn't an issue as such, but it may be prudent to avoid accidentally calling this multiple times. However, if it's a migration fn from old to new storage fields and it's already been called, the old storage would then be empty so another call should be no-op

**Comment by goastler** [comment]
https://substrate.stackexchange.com/questions/9117/how-do-i-upgrade-an-ink-contract-with-storage-changes

**Comment by goastler** [comment]
@forgetso needs to ask hernando about this

**Comment by forgetso** [comment]
Contacted Hernando on Telegram and sent him a link to the question above.

**Comment by forgetso** [comment]
Hernando has nothing for us. Proxy seems easier.
